### ISSUE TITLE
Valid UTF-8 byte sequences in logs are munged.

### ISSUE THREAD

**Issue by yaauie** [cherweg/logstash-input-s3-sns-sqs#40]
In `LogProcessor#process`, we send `String#encode` to each line to convert from a forced `binary` encoding to `utf-8`, replacing each invalid _byte_ with a replacement character.

> ~~~ ruby
>       line = line.encode('UTF-8', 'binary', invalid: :replace, undef: :replace, replace: "\u2370")
> ~~~
> -- [lib/logstash/inputs/s3snssqs/log_processor.rb:29@v2.1.1](https://github.com/cherweg/logstash-input-s3-sns-sqs/blob/v2.1.1/lib/logstash/inputs/s3snssqs/log_processor.rb#L29).

Unfortunately, this causes inputs with valid UTF-8 multi-byte sequences to be munged, since forcing a `binary` source causes the conversion to happen one _byte_ at a time, and the individual bytes of a multibyte sequence are not valid UTF-8 as evidenced by this IRB session:

~~~ ruby
irb(main):001:0> line = '私はガラスを食べられます。それは私を傷つけません'
# => "私はガラスを食べられます。それは私を傷つけません"
irb(main):002:0> line.encoding
# => #<Encoding:UTF-8>
irb(main):003:0> line.valid_encoding?
# => true
irb(main):004:0> line.encode('UTF-8', 'binary', invalid: :replace, undef: :replace, replace: "\u2370")
# => "⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰⍰"
~~~

**Comment by christianherweg0807** [comment]
Hi Ry,
sorry for that. That´s a historic line of code... maybe i could remove it.

**Comment by christianherweg0807** [comment]
This line was originally written, because some methods in ruby die on invalid utf-8 charakters (e.g. split). 
So i tried to encode every line, because i don´t want to loose the whole file, if there is one bad character...
A quick fix could be to test `line.valid_encoding?` before the `line.encode`...

Ideas to make it better?

regards
Christian

**Comment by ryanpersaud** [comment]
It seems like the call to `codec.decode(line)` would throw an exception in many cases if there is an illegal character in the line.  For example if the `json` codec is being used.  Therefore would it make sense to surround the call to `codec.decode(line)` with a begin-rescue-end and make the call to `line.encode()` configurable (defaulting to True for backward compatibility)?

Since an exception thrown by `code.decode(line)` is handled, it should not abort processing of the entire file, and for cases where the codec's decode function does not touch each character in the line, a user has the option of enabling the call to `line.encode()` to ensure that processing does not fail further down the pipeline.

**Comment by christianherweg0807** [comment]
Just released 2.1.2 with a fix for this Problem. Plz let me know if it was helpful.

@ryanpersaud : Thank you for your input

regards
Christian
